testdata/prevote.txt

```diff
@@ -0,0 +1,387 @@
+  Entries:
+  1/11 EntryNormal ""
+  Messages:
+  2->1 MsgAppResp Term:1 Log:0/11
+> 3 handling Ready
+  Ready MustSync=true:
+  Lead:1 State:StateFollower
+  Entries:
+  1/11 EntryNormal ""
+  Messages:
+  3->1 MsgAppResp Term:1 Log:0/11
+> 1 receiving messages
+  2->1 MsgAppResp Term:1 Log:0/11
+  3->1 MsgAppResp Term:1 Log:0/11
+> 1 handling Ready
+  Ready MustSync=false:
+  HardState Term:1 Vote:1 Commit:11
+  CommittedEntries:
+  1/11 EntryNormal ""
+  Messages:
+  1->2 MsgApp Term:1 Log:1/11 Commit:11
+  1->3 MsgApp Term:1 Log:1/11 Commit:11
+> 2 receiving messages
+  1->2 MsgApp Term:1 Log:1/11 Commit:11
+> 3 receiving messages
+  1->3 MsgApp Term:1 Log:1/11 Commit:11
+> 2 handling Ready
+  Ready MustSync=false:
+  HardState Term:1 Vote:1 Commit:11
+  CommittedEntries:
+  1/11 EntryNormal ""
+  Messages:
+  2->1 MsgAppResp Term:1 Log:0/11
+> 3 handling Ready
+  Ready MustSync=false:
+  HardState Term:1 Vote:1 Commit:11
+  CommittedEntries:
+  1/11 EntryNormal ""
+  Messages:
+  3->1 MsgAppResp Term:1 Log:0/11
+> 1 receiving messages
+  2->1 MsgAppResp Term:1 Log:0/11
+  3->1 MsgAppResp Term:1 Log:0/11
+
+# Propose a command on 1 and replicate it to 2.
+propose 1 prop_1
+----
+ok
+
+process-ready 1
+----
+Ready MustSync=true:
+Entries:
+1/12 EntryNormal "prop_1"
+Messages:
+1->2 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "prop_1"]
+1->3 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "prop_1"]
+
+deliver-msgs 2
+----
+1->2 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "prop_1"]
+
+process-ready 2
+----
+Ready MustSync=true:
+Entries:
+1/12 EntryNormal "prop_1"
+Messages:
+2->1 MsgAppResp Term:1 Log:0/12
+
+# 3 is now behind on its log. Attempt to campaign.
+raft-log 3
+----
+1/11 EntryNormal ""
+
+campaign 3
+----
+INFO 3 is starting a new election at term 1
+INFO 3 became pre-candidate at term 1
+INFO 3 [logterm: 1, index: 11] sent MsgPreVote request to 1 at term 1
+INFO 3 [logterm: 1, index: 11] sent MsgPreVote request to 2 at term 1
+
+process-ready 3
+----
+Ready MustSync=false:
+Lead:0 State:StatePreCandidate
+Messages:
+3->1 MsgPreVote Term:2 Log:1/11
+3->2 MsgPreVote Term:2 Log:1/11
+INFO 3 received MsgPreVoteResp from 3 at term 1
+INFO 3 has received 1 MsgPreVoteResp votes and 0 vote rejections
+
+deliver-msgs 1 2
+----
+2->1 MsgAppResp Term:1 Log:0/12
+3->1 MsgPreVote Term:2 Log:1/11
+INFO 1 [logterm: 1, index: 12, vote: 1] rejected MsgPreVote from 3 [logterm: 1, index: 11] at term 1
+3->2 MsgPreVote Term:2 Log:1/11
+INFO 2 [logterm: 1, index: 12, vote: 1] rejected MsgPreVote from 3 [logterm: 1, index: 11] at term 1
+
+# 3 failed to campaign. Let the network stabilize.
+stabilize
+----
+> 1 handling Ready
+  Ready MustSync=false:
+  HardState Term:1 Vote:1 Commit:12
+  CommittedEntries:
+  1/12 EntryNormal "prop_1"
+  Messages:
+  1->2 MsgApp Term:1 Log:1/12 Commit:12
+  1->3 MsgApp Term:1 Log:1/12 Commit:12
+  1->3 MsgPreVoteResp Term:1 Log:0/0 Rejected (Hint: 0)
```

Comment:
Not this PR, just an exploratory question.

I wonder how the `MsgPreVote` sender knows which outbound message this `Resp` corresponds to.

Scenario:
- Node 1 log is outdated. It sends a `MsgPreVote` to node 2.
- Node 1 log is updated, and is now long enough. It sends another `MsgPreVote`.
- Node 2 receives the first message, sends a reject back.
- Node 2 receives the second message, sends an approval back.
- The first (rejected) `Resp` comes back to node 1. Does it consider the first, the second, or both pre-votes rejected?
- If/when the approved `Resp` comes back, how does it identify which pre-vote this corresponds to?

Reply:
I don't think it matters. PreVote is necessarily best-effort, because conditions can chance by the time we hold the actual election. For example, the leader may have woken back up and replicated log entries to the other voters, or a different candidate can also have obtained prevotes and raced with us to win the vote instead. The way to think about it is "if I had held an election right now, would you vote for me?", but it's not a guarantee that we'll actually get the vote, only an indication that we could possibly win an election.

More concretely though, we only record the first vote we receive from a different node:

https://github.com/etcd-io/raft/blob/ffe5efcf2edc7c7236f33eb1f62623c24fc62145/tracker/tracker.go#L261-L264

Once we have a quorum of votes, we'll either hold the election or transition to follower, at which point we clear out our tracked votes:

https://github.com/etcd-io/raft/blob/09c91d8dc0d2f550b69fbe183399e98b1fd6d606/raft.go#L1570-L1584

Also, we won't send multiple `MsgPreVote` in the same precandidate campaign either -- we'll wait for a quorum of prevotes, then transition either to candidate or follower, and then potentially transition back to precandidate and hold another prevote later, at which point we'll send the next `MsgPreVote`. Of course, it's possible that we'll now receive a stale `MsgPreVoteResp` from the previous precandidate campaign if it was delayed while we were transitioning through these states, but such is life.